plugins/flashrom/fu-flashrom-lspcon-i2c-spi-device.c

```diff
@@ -0,0 +1,315 @@
+
+	/* FuFlashromDevice->prepare */
+	if (!FU_DEVICE_CLASS
+	    (fu_flashrom_lspcon_i2c_spi_device_parent_class)->prepare (device,
+								       FWUPD_INSTALL_FLAG_NONE,
+								       error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gint
+fu_flashrom_lspcon_i2c_spi_device_get_bus_number_from_path (const gchar *path,
+							    GError **error)
+{
+	gint bus_number = -1;
+	GRegex *regex = NULL;
+	GMatchInfo *info;
+
+	regex = g_regex_new (I2C_PATH_REGEX, 0, 0, error);
+	if (regex == NULL)
+		return FALSE;
+
+	if (g_regex_match_full (regex, path, -1, 0, 0, &info, error)) {
+		bus_number = g_ascii_strtoll ( g_match_info_fetch (info, 1),
+					       NULL, 10);
+	}
+
+	g_match_info_free (info);
+	g_regex_unref (regex);
+
+	return bus_number;
+}
+
+static gboolean
+fu_flashrom_lspcon_i2c_spi_device_setup (FuDevice *device, GError **error)
+{
+	FuFlashromDevice *parent = FU_FLASHROM_DEVICE (device);
+	FuFlashromLspconI2cSpiDevice *self = FU_FLASHROM_LSPCON_I2C_SPI_DEVICE (device);
+	const gchar *path = NULL;
+	const gchar *hw_id = NULL;
+	g_autofree gchar *vid = NULL;
+	g_autofree gchar *pid = NULL;
+	g_autofree gchar *guid = NULL;
+	g_autofree gchar *contents = NULL;
+	struct flashrom_flashctx *flashctx = NULL;
+	gsize flash_size;
+	uint32_t addr;
+
+	if (g_strcmp0 (fu_flashrom_device_get_programmer_name (parent),
+		       "lspcon_i2c_spi") != 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "invalid programmer");
+		return FALSE;
+	}
+
+	path = fu_udev_device_get_sysfs_path (FU_UDEV_DEVICE (device));
+	self->bus_number = fu_flashrom_lspcon_i2c_spi_device_get_bus_number_from_path (path, error);
+	if (self->bus_number == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s failed to get bus number for " \
+			     "device under %s.",
+			     __func__,
+			     path);
+		return FALSE;
+	}
+
+	hw_id = fu_udev_device_get_sysfs_attr (FU_UDEV_DEVICE (device), "name", error);
+	vid = g_strndup (hw_id, HID_LENGTH / 2);
+	pid = g_strndup (&hw_id[HID_LENGTH / 2], HID_LENGTH / 2);
+	fu_device_add_vendor_id (device, g_strdup_printf ("I2C:%s", vid));
+
+	guid = fwupd_guid_hash_string (g_strdup_printf (DEVICE_GUID_FORMAT, vid, pid));
+	if (guid == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s failed to generate guid.",
+			     __func__);
+		return FALSE;
+	}
+	fu_device_add_guid (device, guid);
+
+	/* FuFlashromLspconI2cSPiDevice->prepare */
+	if (!fu_flashrom_lspcon_i2c_spi_device_prepare (device,
+							FWUPD_INSTALL_FLAG_NONE,
+							error))
+		return FALSE;
+
+	/* set up flashrom layout */
+	flashctx = fu_flashrom_device_get_flashctx (parent);
+	flashrom_layout_set (flashctx, &layout);
+
+	/* read the current flash contents */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+	flash_size = fu_flashrom_device_get_flash_size (parent);
+	contents = g_malloc0 (flash_size);
```

Comment:
I2C isn't that fastest bus.
it seems like you really don't need to read the whole flash, would it be faster to only read up to what data you really need?

Reply:
The amount of data read is limited by the layout set above. I guess I could add some additional regions with only the version area and read those... Let me experiment with that..

Reply:
I mean don't bust your butt over it especially if we're talking milliseconds of difference in probe time.  I just want to avoid a 30s probe that could be a 3s probe.

Reply:
Added additional entries on layout to just read the version section.